nni/algorithms/compression/v2/pytorch/pruning/tools/base.py

```diff
@@ -307,142 +287,128 @@ def calculate_metrics(self, data: Dict) -> Dict[str, Tensor]:
+        If allocator meets an unspecified module name, it will try to use `scalors['_default'][target_name]` to scale its mask.
+        If allocator meets an unspecified target name, it will try to use `scalors[module_name]['_default']` to scale its mask.
+        Passing in a scalor instead of a `dict` of scalors will be treated as passed in `{'_default': {'_default': scalors}}`.
+        Passing in `None` means no need to scale.
     continuous_mask
-        Inherit the mask already in the wrapper if set True.
+        If set True, the part that has been masked will be masked first.
+        If set False, the part that has been masked may be unmasked due to the increase of its corresponding metric.
     """
 
-    def __init__(self, pruner: Pruner, dim: Optional[Union[int, List[int]]] = None,
-                 block_sparse_size: Optional[Union[int, List[int]]] = None, continuous_mask: bool = True):
+    def __init__(self, pruner: Pruner, scalors: Dict[str, Dict[str, Scaling]] | Scaling | None = None, continuous_mask: bool = True):
         self.pruner = pruner
-        self.dim = dim if not isinstance(dim, int) else [dim]
-        self.block_sparse_size = block_sparse_size if not isinstance(block_sparse_size, int) else [block_sparse_size]
-        if self.block_sparse_size is not None:
-            assert all(i >= 1 for i in self.block_sparse_size)
-        elif self.dim is not None:
-            self.block_sparse_size = [1] * len(self.dim)
-        if self.dim is not None:
-            assert all(i >= 0 for i in self.dim)
-            self.dim, self.block_sparse_size = (list(t) for t in zip(*sorted(zip(self.dim, self.block_sparse_size))))  # type: ignore
+        self.scalors: Dict[str, Dict[str, Scaling]] | None = scalors if isinstance(scalors, (dict, type(None))) else {'_default': {'_default': scalors}}  # type: ignore
         self.continuous_mask = continuous_mask
 
-    def generate_sparsity(self, metrics: Dict) -> Dict[str, Dict[str, Tensor]]:
+    def _get_scalor(self, module_name: str, target_name: str) -> Scaling | None:
+        # Get scalor for the specific target in the specific module. Return None if don't find it.
+        # `module_name` is not used in current nni version, will support different modules using different scalors in the future.
+        if self.scalors:
+            default_module_scalors = self.scalors.get('_default', {})
+            default_target_scalor = default_module_scalors.get(target_name, default_module_scalors.get('_default', None))
+            module_scalors = self.scalors.get(module_name, {})
+            return module_scalors.get(target_name, module_scalors.get('_default', default_target_scalor))
+        else:
+            return None
+
+    def _expand_mask(self, module_name: str, target_name: str, mask: Tensor) -> Tensor:
+        # Expand the shrinked mask to the pruning target size.
+        scalor = self._get_scalor(module_name=module_name, target_name=target_name)
```

Comment:
you see you have to use`module_name` and `target_name` to index scalors, and you also use the same index to retrieve the mask from wrapper. it may imply that scalors should be stored in wrapper.

Reply:
yes, this is our final goal, but not in this pr